elastic/kibana
# [Security solution] Bedrock streaming and token tracking

## Summary

Part 2 of 3 of Streaming in Security AI Assistant ([Part 1](https://github.com/elastic/kibana/pull/169729))

1. Implement streaming for the Bedrock connector
2. Add `invokeStream` subaction to OpenAI and Bedrock. Handles standardizing our assistant stream request to OpenAI and Bedrock formats, and parses the response from the OpenAI and Bedrock formats to a simple string for the consumer to stream
3. Adds token tracking for Bedrock, streaming (`invokeStream`) and non streaming (`run`, `test`), in action_executor.ts
    - Note: the `invokeAI` method of Bedrock does not have token tracking. I plan remove that method in part 3 of 3 of the streaming work. At that time I will add a link to the token tracking dashboard form the Bedrock connector edit, like we [have for OpenAI](https://github.com/elastic/kibana/pull/159075)

### package.json changes
I could not find documentation on how to handle the Bedrock codec. I did find awsdocs/aws-doc-sdk-examples#5541 that we can track for when documentation should be added.

I followed the bedrock extension in the langchainjs repo here: https://github.com/langchain-ai/langchainjs/pull/2367/files#diff-ead6be41360b3bf1c4559ea4d0069a015b69a57d5f8c51d0cce73253fde30aa2R246-R268. This is how I found that we'd need these packages: `@smithy/eventstream-codec`, `@smithy/util-utf`

### To test:

1. Create Bedrock connector with [Bedrock credentials](https://p.elstc.co/paste/rxa+OUI2#Nzv+-vFnsJB7xIkzb18rdVgNAJVTmYLZHsrKcHmt02N) and assign to your security AI assistant conversation
2. Ensure langchain is not enabled. Knowledge base should be off:
<img width="767" alt="Screenshot 2023-11-14 at 10 20 54 AM" src="https://github.com/elastic/kibana/assets/6935300/45fc9d15-71b3-4682-911d-a706492c9eac">

3. Send message to the connector. Test that:
    - response streams
    - "Stop generating" button works
    - "Regenerate" button works
4. Save [this file](https://p.elstc.co/paste/mYL-wztM#vJA0JR9povFgfD122VZEav5sj2R+YqOp8ZFOHA43ICJ) as `export.ndjson` and upload to your saved objects
5. Click into the saved search we just uploaded, "Bedrock token tracking"
6. Ensure you see events with tokens for bedrock
7. Back to the Assistant, create a bedrock connector with bad credentials and ensure the Assistant returns an error when you attempt to use this connector


### A note on `api_alerting_integration` tests

I spent 2 full work days trying to get these working. I’m trying to get the stream response from `supertest`. I can see the stream running on the backend, but when it comes back to `supertest` it only returns the `readableState`, not the stream itself.  If there is anyone who can pair, I’d appreciate it. Otherwise, the code is covered in jest by `x-pack/plugins/stack_connectors/server/connector_types/bedrock/bedrock.test.ts`
